---
title: "ggbio: Biological data plots"
subtitle: "SISBID 2016"
author: "Di Cook (dicook@monash.edu, @visnut); Heike Hofmann (heike.hofmann@gmail.com, @heike_hh)"
date: "`r Sys.Date()`"
output:
  ioslides_presentation:
    transition: default
    widescreen: true
css:
  styles.css
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  collapse = TRUE,
  comment = "#>",
  fig.height = 4,
  fig.width = 8,
  fig.align = "center",
  cache = FALSE
)
```


```{r, echo = FALSE}
library(ggplot2)
library(tidyr)
library(dplyr)
library(lubridate)
library(scales)
library(readr)
library(ggbio)
library(biovizBase)
library(GenomicRanges)
```

## What is ggbio?

This is a package originally developed by Tengfei Yin and Michael Lawrence building on ggplot2, making it easier to produce some standard genomic plots. 

- ideogram
- tracks
- karyogram
- manhattan plot
- circular layout

##

```{r fig.show='hold'}
p.ideo <- Ideogram(genome = "hg19")
p.ideo
```

##

```{r}
p.ideo + xlim(GRanges("chr2", IRanges(1e8, 1e8+10000)))
```

##

```{r}
Ideogram(genome = "hg19", xlabel = TRUE)
```

## Tracks

Lining up multiple plots along the chromosome is an important part of plotting genomic data

## 

```{r}
library(Homo.sapiens)
class(Homo.sapiens)
##
data(genesymbol, package = "biovizBase")
wh <- genesymbol[c("BRCA1", "NBR1")]
wh <- range(wh, ignore.strand = TRUE)
p.txdb <- autoplot(Homo.sapiens, which = wh)
p.txdb
```

##

```{r}
autoplot(Homo.sapiens, which = wh, label.color = "black", color = "brown",
fill = "brown")
```

##

```{r}
autoplot(Homo.sapiens, which = wh, gap.geom = "chevron")
```

##

```{r}
autoplot(Homo.sapiens, which = wh, stat = "reduce")
```

##

```{r}
columns(Homo.sapiens)
autoplot(Homo.sapiens, which = wh, columns = c("TXNAME", "GO"), names.expr = "TXNAME::GO")
```

##

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
autoplot(txdb, which = wh)
```

##

```{r}
library(EnsDb.Hsapiens.v75)
ensdb <- EnsDb.Hsapiens.v75
autoplot(ensdb, GenenameFilter("PHKG2"))
```

##

Alternatively,

```{r eval=FALSE}
gr <- GRanges(seqnames=16, IRanges(30768000, 30770000), strand="*")
autoplot(ensdb, GRangesFilter(gr, "overlapping"), names.expr="gene_name")
```

or

```{r eval=FALSE}
autoplot(ensdb, GeneidFilter(c("ENSG00000196118", "ENSG00000156873")))
```

##

```{r}
gr.txdb <- crunch(txdb, which = wh)
## change column to model

colnames(values(gr.txdb))[4] <- "model"
grl <- split(gr.txdb, gr.txdb$tx_id)
## fake some random names
names(grl) <- sample(LETTERS, size = length(grl), replace = TRUE)
grl
```

##

```{r}
autoplot(grl, aes(type = model))
ggplot() + geom_alignment(grl, type = "model")
```

##

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
bg <- BSgenome.Hsapiens.UCSC.hg19
p.bg <- autoplot(bg, which = wh)
## no geom
p.bg
## segment
p.bg + zoom(1/100)
## rectangle
p.bg + zoom(1/1000)
## text
p.bg + zoom(1/2500)
```

##

```{r}
bg <- BSgenome.Hsapiens.UCSC.hg19
## force to use geom segment at this level
autoplot(bg, which = resize(wh, width = width(wh)/2000), geom = "segment")
```

##

```{r}
fl.bam <- system.file("extdata", "wg-brca1.sorted.bam", package = "biovizBase")
wh <- keepSeqlevels(wh, "chr17")
autoplot(fl.bam, which = wh)
```

##

```{r}
fl.bam <- system.file("extdata", "wg-brca1.sorted.bam", package = "biovizBase")
wh <- keepSeqlevels(wh, "chr17")
autoplot(fl.bam, which = resize(wh, width = width(wh)/10), geom = "gapped.pair")
```

##

```{r}
bg <- BSgenome.Hsapiens.UCSC.hg19
p.mis <- autoplot(fl.bam, bsgenome = bg, which = wh, stat = "mismatch")
p.mis
```

##

```{r}
p.ideo <- Ideogram(genome = "hg19")
gr17 <- GRanges("chr17", IRanges(41234415, 41234569))
tks <- tracks(p.ideo, mismatch = p.mis, ref = p.bg, gene = p.txdb,
  heights = c(2, 3, 1, 4)) + xlim(gr17) + theme_tracks_sunset()
tks
```

##

```{r fig.show='hold'}
data("CRC", package = "biovizBase")
head(hg19sub)
autoplot(hg19sub, layout = "circle", fill = "gray70")
p <- ggbio() + circle(hg19sub, geom = "ideo", fill = "gray70") +
circle(hg19sub, geom = "scale", size = 2) +
circle(hg19sub, geom = "text", aes(label = seqnames), vjust = 0, size = 3)
p
p <- ggbio(trackWidth = 10, buffer = 0, radius = 10) + circle(hg19sub, geom = "ideo", fill = "gray70") +
circle(hg19sub, geom = "scale", size = 2) +
circle(hg19sub, geom = "text", aes(label = seqnames), vjust = 0, size = 3)
p
head(mut.gr)
p <- ggbio() + circle(mut.gr, geom = "rect", color = "steelblue") +
circle(hg19sub, geom = "ideo", fill = "gray70") +
circle(hg19sub, geom = "scale", size = 2) +
circle(hg19sub, geom = "text", aes(label = seqnames), vjust = 0, size = 3)
p
```

## Resources

- Yin T, Cook D and Lawrence M (2012). “ggbio: an R package for extending the grammar of graphics for genomic data.” Genome Biology, 13(8), pp. R77.
- [bioconductor page](https://bioconductor.org/packages/release/bioc/html/ggbio.html)

## Share and share alike

This work is licensed under the Creative Commons Attribution-Noncommercial 3.0 United States License. To view a copy of this license, visit http://creativecommons.org/licenses/by-nc/ 3.0/us/ or send a letter to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.
